docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,526 @@
+## Introduction
+
+This proposal discusses adding the support for referencing an ACK custom resource
+in another custom resource. The reconciler will wait for the referenced resource
+to get created first and then create the custom resource.
+
+For example, currently if ACK customer needs to create an APIGatewayv2 integration,
+they first need to
+1. create an `Api` custom resource
+2. wait for `Api` resource creation to finish
+3. manually find the `ApiId` from `Status` of `Api` resource
+4. create an `Integration` resource with the `ApiId`
+
+Having to manually wait for the resource creation makes it impossible to create
+ACK resources using any gitops method where all the custom resource manifest
+can be managed in the source control and applied together to k8s cluster.
+
+With this approach of referenced resources, customers will be able to reference
+Api resource in Integration resource and apply both resource manifests
+together in their k8s cluster. This will allow customers to manage ACK custom
+resources using gitops methodologies. ACK controller will wait for API resource
+to get created and then use the ApiId to create Integration resource.
+
+The deletion strategy for the referenced resource is out of scope for this proposal
+right now. For Ex: If API resource is referenced in integration resource, the deletion
+of API resouorce will not be blocked. We plan to handle deletion strategy in next
+revision.
+
+## Scope
+
+### First Revision
+* Allow referencing resource from same api group in custom resource definitions
+* Allow referencing list of resources from same api group in custom resource
+definitions
+
+### Future Improvements
+* Add support for custom hooks while dereferencing resources
+* Add support for `selector` for referencing the objects
+* Add support for referencing resource from different api-groups
+* Add support for not deleting a resource if it is referenced in another resource
+* Add support for finite retries and terminal conditions when resolving references
+
+## Preferred Approach
+
+### High Level Overview
+
+* Introduce new construct in generator.yaml which will indicate to code-generator
+if a new spec field with type `AWSResourceReference` should be added to the custom
+resource definition.
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
```

Comment:
Not sure if i understand this statement. Why can't we have spec fields and reference at same time?

Reply:
For ex: Why would one want to provide an `ApiId` and API reference at the same time?
If we allow this, it will make us choose which one to prioritize etc. I thought it's better to not go down that road.